picasso/src/main/java/com/squareup/picasso3/BitmapHunter.kt

```diff
@@ -0,0 +1,350 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.picasso3
+
+import android.net.NetworkInfo
+import com.squareup.picasso3.MemoryPolicy.Companion.shouldReadFromMemoryCache
+import com.squareup.picasso3.Picasso.LoadedFrom
+import java.io.IOException
+import java.io.InterruptedIOException
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.Future
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicReference
+
+internal class BitmapHunter(
+  val picasso: Picasso,
+  val key: String,
+  var data: Request,
+  private val dispatcher: Dispatcher,
+  private val cache: PlatformLruCache,
+  private val requestHandler: RequestHandler,
+  private var retryCount: Int,
+  action: Action?,
+  priority: Picasso.Priority
+) : Runnable {
+  companion object {
+    val NAME_BUILDER: ThreadLocal<StringBuilder> = object : ThreadLocal<StringBuilder>() {
+      override fun initialValue(): StringBuilder = StringBuilder(Utils.THREAD_PREFIX)
+    }
+    val SEQUENCE_GENERATOR = AtomicInteger()
+    val ERRORING_HANDLER: RequestHandler = object : RequestHandler() {
+      override fun canHandleRequest(data: Request): Boolean = true
+
+      override fun load(picasso: Picasso, request: Request, callback: Callback) {
+        callback.onError(IllegalStateException("Unrecognized type of request: $request"))
+      }
+    }
+
+    fun forRequest(
+      picasso: Picasso,
+      dispatcher: Dispatcher,
+      cache: PlatformLruCache,
+      action: Action
+    ): BitmapHunter {
+      val request = action.request
+      val requestHandlers = picasso.getRequestHandlers()
+
+      // Index-based loop to avoid allocating an iterator.
+      for (i in 0..requestHandlers.lastIndex) {
+        val requestHandler = requestHandlers[i]
+        if (requestHandler.canHandleRequest(request)) {
+          return BitmapHunter(
+            picasso = picasso,
+            key = action.request.key,
+            data = action.request,
+            dispatcher = dispatcher,
+            cache = cache,
+            requestHandler = requestHandler,
+            retryCount = requestHandler.retryCount,
+            action = action,
+            priority = action.request.priority
+          )
+        }
+      }
+
+      return BitmapHunter(
+        picasso = picasso,
+        key = action.request.key,
+        data = action.request,
+        dispatcher = dispatcher,
+        cache = cache,
+        requestHandler = ERRORING_HANDLER,
+        retryCount = ERRORING_HANDLER.retryCount,
+        action = action,
+        priority = action.request.priority
```

Comment:
adding `key`, `data`, `retryCount`, `priority` to the constructor burdens the callers.  Can we move those back to internal fields and initialize at the declaration sites?